java/com/google/re2j/Unicode.java

```diff
@@ -227,6 +229,99 @@ static int simpleFold(int r) {
     return toUpper(r);
   }
 
+  static int[] optimizedFoldOrbit(int r) {
+    if (r >= ORBIT_MIN_VALUE) {
+      return FOLD_MAP.get(r);
+    } else return null;
+  }
+
+  static int optimizedFoldNonOrbit(int r) {
+    int l = toLower(r);
+    if (l != r) {
+      return l;
+    }
+    return toUpper(r);
+  }
+  
+  private static final FoldMap FOLD_MAP = new FoldMap(UnicodeTables.CASE_ORBIT.length * 4);
```

Comment:
Whence 4?  Is it a load factor based on MAX_DISTANCE?

I don't think this should be a parameter.  FoldMap is instantiated exactly once; specialize it.